File: api/src/main/java/com/bindglam/goldengine/lang/LanguageComponent.java
package com.bindglam.goldengine.lang;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;

public final class LanguageComponent {
    private final String raw;

    public LanguageComponent(String raw) {
        this.raw = raw;
    }

    public Component component(Object... args) {
        int argIdx = 0;

        String result = raw;
        while(result.contains("%s")) {
            result = result.replaceFirst("%s", args[argIdx++].toString());
        }
        return MiniMessage.miniMessage().deserialize(result);
    }
}

File: core/src/main/kotlin/com/bindglam/goldengine/language/LanguageComponent.kt
package com.bindglam.goldengine.language

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage

class LanguageComponent(private val raw: String) {
    fun component(vararg args: Any): Component {
        var argIdx = 0

        var result = raw
        while (result.contains("%s")) {
            result = result.replaceFirst("%s".toRegex(), args[argIdx++].toString())
        }
        return MiniMessage.miniMessage().deserialize(result)
    }
}